File: src/main/java/com/polygloat/dtos/request/EditApiKeyDTO.java
package com.polygloat.dtos.request;

import com.fasterxml.jackson.annotation.JsonGetter;
import com.fasterxml.jackson.annotation.JsonSetter;
import com.polygloat.constants.ApiScope;
import lombok.AllArgsConstructor;
import lombok.Builder;
import lombok.Data;
import lombok.NoArgsConstructor;

import javax.validation.constraints.NotEmpty;
import javax.validation.constraints.NotNull;
import java.util.Set;
import java.util.stream.Collectors;

@Data
@Builder
@NoArgsConstructor
@AllArgsConstructor
public class EditApiKeyDTO {
    @NotNull
    private Long id;

    @NotEmpty
    private Set<ApiScope> scopes;

    @JsonSetter("scopes")
    public void jsonSetScopes(Set<String> scopes) {
        this.scopes = scopes.stream().map(ApiScope::fromValue).collect(Collectors.toSet());
    }

    @JsonGetter("scopes")
    public Set<String> jsonGetScopes() {
        return this.scopes.stream().map(ApiScope::getValue).collect(Collectors.toSet());
    }
}

File: backend/data/src/main/kotlin/io/tolgee/dtos/request/apiKey/EditApiKeyDTO.kt
package io.tolgee.dtos.request.apiKey

import com.fasterxml.jackson.annotation.JsonGetter
import com.fasterxml.jackson.annotation.JsonSetter
import io.tolgee.model.enums.ApiScope
import java.util.stream.Collectors
import javax.validation.constraints.NotEmpty
import javax.validation.constraints.NotNull

data class EditApiKeyDTO(
  @field:NotNull
  var id: Long = 0,
  @field:NotEmpty
  var scopes: Set<ApiScope> = setOf()
) {
  @Suppress("unused")
  @JsonSetter("scopes")
  fun jsonSetScopes(scopes: Set<String>) {
    this.scopes = scopes.map { value -> ApiScope.fromValue(value) }.toSet()
  }

  @Suppress("unused")
  @JsonGetter("scopes")
  fun jsonGetScopes(): Set<String> {
    return scopes.stream().map { obj: ApiScope -> obj.value }.collect(Collectors.toSet())
  }
}
